### DISCUSSION TITLE
Extra data column on users tabel (not row_user_extra_data).

### DISCUSSION THREAD

**Discussion by mrc-bsllt** [supabase/supabase#19942] [Questions] [answered]
Is it possible to add additional fields to the users table? 
I know I could add them in row_user_meta_data, but what if I wanted to add 'username' for example, and make it isUnique? 
If I have to fill in username, email and password in a form I need to be able to block the creation of a user if the username has already been taken.
It would be nice to be able to add extra rows to the user table.

**Comment by GaryAustin1** [comment, marked as answer]
No you cannot modify the auth schema.  Supabase could reset your table at anytime with column adjustments.
This is normally done use a public.profiles table and a trigger on auth.users insert to populate it when a user is created.

**Reply by mrc-bsllt** [reply]
Thank you for your reply.
However, this means that when I create a new user and then save additional data in the profiles table, in case there is a duplicate username, I have to delete the user, right? Or are there better solutions?

**Reply by GaryAustin1** [reply]
I don't understand your question.  Normally the profile table is a 1 to 1 match, sometimes using a foreign key back to auth.users table UUID as the primary key.
You then update your profiles table later with changes you want.
https://supabase.com/docs/guides/auth/managing-user-data covers the basic ideas.

**Reply by GaryAustin1** [reply]
Ah, I see on your unique question.  You would in the insert trigger need to test for that and decide what to do.  Add a number to it or raise an exception and abort the user signup totally.  Not sure how having in auth users table would make that any different if it is a unique column.  If your trigger fails, the entire auth operation fails back to the API. 
Edit:  Note that if your insert to the profiles table fails because of a unique violation then the overall operation fails.  So if you are wanting to just fail, you don't need to test before the insert... just let it fail.

**Reply by mrc-bsllt** [reply]
Great! Thanks again for your reply!

**Reply by mrc-bsllt** [reply]
@GaryAustin1 Sorry to come back to this topic, but I am a little confused. How do I check that the username has not already been used before I complete the whole signup process, and can then block it?
The profiles table only needs to be accessible to a specific user, so I have to have created it first. I hope I have been clear in my explanation.

**Reply by GaryAustin1** [reply]
Well you originally implied you just wanted the signup to fail.  If you try the insert in the user management style trigger it will fail if duplicate and then fail back to the auth insert, which will then fail the signup.  Not greatest user experience.

In triggers you can read tables so you could check the profile table for that username with a select and do what you want if it exists.  The user management trigger functions as shown in the guide are 'security definer' type so bypass any RLS.

If you want to from your app see if it exists you would write a security definer function and call it with rpc.   Provide the proposed user name, call the rpc with that as a parameter and return true/false if it exists.
